src/QUICConnection.ts

```diff
@@ -410,73 +339,180 @@ class QUICConnection extends EventTarget {
+      await this.send();
+    } else if (this.type === 'server') {
+      if (data == null || remoteInfo == null) {
+        throw new errors.ErrorQUICConnectionStartData(
+          'Starting a server connection requires initial data and remote information'
+        );
+      }
+      // This chain up recv and send and setup the max idle timeout
+      await this.recv(data, remoteInfo);
+    }
+    try {
+      // This will block until the connection is established
+      // Which also depends on a mutual TLS handshake
+      // It is expected that multiple `recv` and `send` pairs
+      // will be called to complete the connection establishment
+      await Promise.race([
+        this.secureEstablishedP, // This might reject with a relevant error
+        abortP, // This might abort for some other reason!
+      ]);
+    } catch (e) {
+      let e_ = e;
+      // This should only be true if we are infact aborted due to start timeout
+      if (ctx.signal.aborted) {
+        // No `QUICStream` objects could have been created, however quiche stream
+        // state should be cleaned up, and this can be done synchronously
+        for (const streamId of this.conn.readable() as Iterable<StreamId>) {
+          this.conn.streamShutdown(streamId, quiche.Shutdown.Read, 0);
+        }
+        for (const streamId of this.conn.writable() as Iterable<StreamId>) {
+          this.conn.streamShutdown(streamId, quiche.Shutdown.Write, 0);
+        }
+        // According to RFC9000, closing while in the middle of a handshake
+        // should use a transport error code `APPLICATION_ERROR`.
+        // For this library we extend this "handshake" phase to include the
+        // the TLS handshake too.
+        // This is also the behaviour of quiche when the connection is not
+        // in a "safe" state to send application errors (where `app` parameter is true).
+        // https://www.rfc-editor.org/rfc/rfc9000.html#section-10.2.3-3
+        this.conn.close(
+          false,
+          ConnectionErrorCode.ApplicationError,
+          Buffer.from('')
+        );
+        const localError = this.conn.localError()!;
+        e_ = new errors.ErrorQUICConnectionLocal(
+          'Failed to start QUIC connection due to start timeout',
+          {
+            data: localError,
+            cause: e
+          }
+        );
+        this.dispatchEvent(
+          new events.EventQUICConnectionError({
+            detail: e_
+          })
+        );
+        this.dispatchEvent(
+          new events.EventQUICConnectionClose({
+            detail: {
+              type: 'local',
+              ...localError
+            }
+          })
+        );
+      }
+      // Wait for the connection to be fully closed
+      // It is expected that max idle timer will eventually resolve this
+      await this.closedP;
+      // Throw the augmented exception if it is augmented
+      // Otherwise throw the original
+      throw e_;
```

Comment:
This is throwing the `ErrorQUICConnectionLocal` with the original error as the cause. It makes more sense to throw the original error here rather than the derived one. the `ErrorQUICConnectionLocal` still makes sense for dispatching the error event.

In terms of usage, if the connection or client creation timed out due to timeout or ctx signal abortion. I'd rather get their specific errors thrown back up.

Reply:
Ok this is fine. Already done in ed97908f546872d61b33dc7e874f5c3437bf8acc